Framework/API/src/FileProperty.cpp

```diff
@@ -396,5 +398,44 @@ std::string FileProperty::convertExtension(const std::string &filepath) const {
   fullpath.setExtension(ext);
   return fullpath.toString();
 }
+
+/** Expand user variables in file path.
+ *  On Windows and UNIX, ~ is replaced by the user's home directory, if found.
+ *  If the path contains no user variables, or expansion fails, the path is
+ *  returned unchanged, for errors to be dealt with by the calling function.
+ *  Note: this function does not support the "~user/blah" format for a named
+ *  user's home directory - if this is encountered, the filepath is returned
+ *  unchanged.
+ *  @param filepath The path to expand
+ *  @return The expanded path
+ */
+std::string FileProperty::expandUser(const std::string &filepath) const {
+  auto start = filepath.begin();
+
+  if (*start != '~') // No user variables in filepath
+    return filepath;
+
+  // Position of the first slash after the variable
+  auto nextSlash = find_if(start, filepath.end(), &isSlash);
+
+  // ~user/blah format - no support for this as yet
+  if (std::distance(start, nextSlash) != 1)
+    return filepath;
+
+  char *home;
```

Comment:
I'm finding this a little hard to parse. I'm also thinking that, for a given process, the value of `HOME`/`USERPROFILE` could probably be cached once and then reused.

If we put a function at anonymous namespace scope, something like:
```cpp
const std::string & getHomePath() {
    static std::string homePath;
    static bool initialized(false);
    if(initialized) return homePath;
    initialized = true;
    homePath = std::getenv("HOME");
    if(!homePath.empty()) return homePath;
    homePath = std::getenv("USERPROFILE");
    return homePath;
}
```
then the environment lookup will happen once and it'll be a little clearer.